onnxruntime/core/optimizer/compute_optimizer/upstream_reshape_actors.cc

```diff
@@ -0,0 +1,283 @@
+    const logging::Logger& logger,
+    std::vector<int>& propagate_input_indices,
+    std::unordered_map<int, std::vector<DimCompareRet>>& all_input_cmp_rets,
+    std::function<void(Node& node)>& shape_update_func) {
+  LOG_DEBUG_INFO(logger, "Enter MatMulReshapeActor::PreCheck for node " + current_node.Name());
+
+  int current_node_output_index = info.GetDataProducerOutputIndex();
+  const NodeArg* data_input_arg = current_node.OutputDefs()[current_node_output_index];
+
+  propagate_input_indices.clear();
+  all_input_cmp_rets.clear();
+
+  propagate_input_indices.reserve(current_node.InputDefs().size());
+
+  // Check limited input shape combinations, supported cases:
+  // 1. The first input have rank = 3, second input has rank = 2.
+  //     Example: [2, 16, 1024] * [1024, 1024], having output shape [2, 16, 1024], merging the first two dims.
+
+  auto lhs_input_shape = current_node.InputDefs()[0]->Shape();
+  auto rhs_input_shape = current_node.InputDefs()[1]->Shape();
+
+  auto lhs_input_rank = lhs_input_shape->dim_size();
+  auto rhs_input_rank = rhs_input_shape->dim_size();
+  if (lhs_input_rank != 3 || rhs_input_rank != 2) {
+    LOG_DEBUG_INFO(logger, "Fail MatMulReshapeActor::PreCheck for node " + current_node.Name() +
+                               ": lhs_input_rank is " + std::to_string(lhs_input_rank) +
+                               ", rhs_input_rank is " + std::to_string(rhs_input_rank));
+    return false;
+  }
+
+  auto [success, ret] = CompareInputShapeWithOutputShape(data_input_arg->Shape(),
+                                                         lhs_input_shape);
+
+  if (!success) {
+    LOG_DEBUG_INFO(logger, "Fail MatMulReshapeActor::PreCheck for node " + current_node.Name() +
+                               ": reshape's data input rank < passthrough node's input rank");
+    return false;
+  }
+
+  all_input_cmp_rets[0] = ret;
+
+  propagate_input_indices.push_back(0);
+  shape_update_func = [&info](Node& node) -> void {
+    ORT_ENFORCE(static_cast<size_t>(1) == node.MutableOutputDefs().size());
+    node.MutableOutputDefs()[0]->SetShape(
+        CreateNewShapeWithMergedTwoLeadingDims(node.MutableOutputDefs()[0]->Shape(), info.last_dim));
+  };
+
+  return propagate_input_indices.size() > 0;
+}
+
+bool LayerNormalizationReshapeActor::PreCheck(
+    const Node& current_node, const ReshapeInfo& info,
+    const logging::Logger& logger,
+    std::vector<int>& propagate_input_indices,
+    std::unordered_map<int, std::vector<DimCompareRet>>& all_input_cmp_rets,
+    std::function<void(Node& node)>& shape_update_func) {
+  LOG_DEBUG_INFO(logger, "Enter LayerNormalizationReshapeActor::PreCheck for node " + current_node.Name());
+
+  auto axis = static_cast<int64_t>(current_node.GetAttributes().at("axis").i());
+  axis = axis < 0 ? axis + current_node.InputDefs()[0]->Shape()->dim_size() : axis;
+
+  // Make sure the layer norm's reduction happens after the axis we want to slice.
```

Comment:
May we list the criterion for layernotm `PreCheck` like we did for the other two actors (numerically)?